design/20220722-publicbundle.md

```diff
@@ -0,0 +1,289 @@
+	"/etc/pki/tls/cacert.pem",                           // OpenELEC
+	"/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem", // CentOS/RHEL 7
+	"/etc/ssl/cert.pem",                                 // Alpine Linux
+}
+```
+
+It's a chore to manually specify a location for a trust store, and it's unfamiliar and error-prone
+for many users. However, if a user instead overwrites any of these files with their own
+`trust`-generated certificate trust bundle they'll currently overwrite their public trust bundle
+provided by their OS / container base image.
+
+Unless the user's application only talks to services which the user also controls, what users are
+likely to really want in most cases is a bundle containing both the publicly trusted certs _and_
+the user's own roots.
+
+It's possible in most distros to append to the cert bundle, but that usually involves running some
+command to generate the final bundle, which would change how the user invokes their application
+inside their container.
+
+Ideally, the whole bundle a user requires would simply be available at one of the above locations.
+
+### Goals
+
+- A user configuring a `Bundle` can add a public trust bundle at least as easily as they can add certs from a `ConfigMap` or `Secret`
+- It's clear to users exactly what version of the public bundle is used at any given time, for auditing purposes
+- A user whose application must use both public and private CAs can confidently overwrite their "system" trust store in a container using a trust `Bundle`
+- A user can update any bundle in use with any version of the trust controller and be confident that it will work
+
+### Non-Goals
+
+- Editing/pruning the public trust bundle (out of scope for the initial feature but might be desirable later)
+- Supporting [OS-specific](https://superuser.com/questions/437330/how-do-you-add-a-certificate-authority-ca-to-ubuntu) certificate update mechanisms
+- Supporting a plethora of options for sources of public trust bundles; we might want this in the future but we should start with one
+
+## Proposal
+
+We propose that the cert-manager project becomes a distributor of its own public trust bundle, which at
+least to start will be a carbon copy of the Mozilla trust bundle.
+
+That bundle in will in turn be packaged into a container which will be used as a sidecar in the trust
+controller's pod, with the ability to present its own bundle to the controller container.
+
+A deployed instance of trust will therefore contain two containers in a pod; the trust controller itself, and a bundle
+container with which trust will communicate in order to fetch a trust bundle.
+
+We'll also add a `publicBundle` field as a source, which is the method by which users include the bundle from the bundle container.
+
+```yaml
+apiVersion: trust.cert-manager.io/v1alpha1
+kind: Bundle
+metadata:
+  name: my-org.com
+spec:
+  sources:
+  - configMap:
+      name: "my-root-cm"
+      key: "root-cert.pem"
+  - publicBundle:
```

Comment:
It may be the case that in future that it is not only publicly trusted bundles which are in the sidecar, but other ones as well. For that reason, I think it makes sense to rename this field to something like `sideCar`, `wellKnown` or `imported` (just some ideas).

Reply:
I think you're maybe envisioning a much bigger and more complex role for the sidecar than I've designed here.

What other bundles would you expect to ship? If we wanted other bundles, could we not ship other sidecars rather than creating one big sidecar?

I think `publicBundle` or something that explicitly references the fact that the bundle is publicly trusted is a big UX win. The action the end user wants is to include a bundle of publicly trusted certificates in their trust bundle - i.e. I think that in a `Bundle` resource they should be specifying what they want, not where it comes from.

Reply:
Yes, regardless of what the end implementation we choose or the scope, if we intend people to bring their own sidecars containing their corps CA or “public certs but not those signed by the wicked witch”, `publicBundle` is the wrong name for that source, and we should not have to update the trust.cert-manager.io API every time a user wants a specific home grown bundle.

> i.e. I think that in a Bundle resource they should be specifying what they want, not where it comes from.

I disagree here- the source type _is_ the where (or domain) you want to get it from. The keys or names referenced on the type is the what.

Reply:
I'm open to changing; I'd not really thought about people wanting to embed their own sidecars. We _could_ support that, but I'm not really sure why someone would want to do that instead of using a ConfigMap or Secret like we already support. A public trust bundle is a special case IMO.

I went with "packagedBundle" in #46 and kinda half-assumed that people might package their own sidecars.

Reply:
On the surface this sounds to me like we are creating a 'filesystem' source and basically shipping a container containing a trust bundle. I don't see why this sidecar mechanism is specifically related to 'public' bundles. If we are fixing the list of CAs included in here/not making it extensible, I also think it makes more sense to store them in the same container as the binary anyway (what happens if someone runs the controller _without_ the sidecar?).

What's the difference between this and us creating a ConfigMap containing the public CAs that we want to distribute and configuring it by default as a source?

Reply:
>  I also think it makes more sense to store them in the same container as the binary anyway

The reason not to do this is talked about at some length in the design document already.

>  (what happens if someone runs the controller without the sidecar?).

Either we refuse to start the controller or they simply don't get to use a sidecar bundle.

> What's the difference between this and us creating a ConfigMap containing the public CAs that we want to distribute and configuring it by default as a source?

The main difference really is the configmap was my original suggestion which was rejected in the biweekly meeting I presented it at, and this is the compromise that was agreed upon.